File: src-ts/Commands/Entertainment/meme.ts
import { APIEmbed } from "discord.js";
import { Command } from "../../Interfaces";
import fetch from "node-fetch";

export const command: Command = {
  name: "meme",
  description: "Sends random meme from r/memes",
  category: "Entertainment",
  aliases: [],
  run: async (variables) => {
    const link = "https://www.reddit.com/r/memes/random/.json";

    fetch(link).then((response) => {
      return response.json();
    }).then((response: any) => {
      const post = response[0].data.children[0].data;

      const url   = "https://www.reddit.com" + post.permalink;
      const title = post.title;
      const image = post.url_overridden_by_dest;

      const votes    = post.ups;
      const comments = post.num_comments;

      const embed: APIEmbed = {
        color: 0xe0af68,
        title,
        url,
        image: {url: image},
        footer: {text: `👍 ${votes} | 💬 ${comments}`}
      };

      variables.client.messager.send_embed(variables, embed);
    });
  }
};

File: src/bot/commands/entertainment/meme.rs
use tracing::error;

use crate::{
    bot::commands::{Context, Error},
    request::reddit_structs::RedditPost,
};

/// Sends random meme from r/memes
#[poise::command(slash_command, prefix_command, category = "Entertainment")]
pub async fn meme(ctx: Context<'_>) -> Result<(), Error> {
    let link = "https://meme-api.com/gimme/memes";
    let Ok(url) = reqwest::Url::parse(link) else {
        error!("Couldn't parse the URL.");
        message!(error, ctx, ("An error occured, please try again later."); false);
        return Ok(());
    };

    let Ok(res) = ctx.data().reqwest_client.get(url).send().await else {
        error!("Couldn't fetch from: {}", link);
        message!(error, ctx, ("An error occured, please try again later."); false);

        return Ok(());
    };

    let Ok(res_str) = res.text().await else {
        error!("Couldn't get respoense.");
        message!(error, ctx, ("An error occured, please try again later."); false);
        return Ok(());
    };

    let Ok(res_last) = sonic_rs::from_str::<RedditPost>(&res_str) else {
        error!("Couldn't serialize the data. Link: {}", link);
        message!(error, ctx, ("An error occured, please try again later."); false);

        return Ok(());
    };

    message!(
        embed,
        ctx,
        vec![serenity::builder::CreateEmbed::new()
            .color(0xE0AF68)
            .title(res_last.title)
            .url(res_last.post_link)
            .image(res_last.url)
            .footer(serenity::builder::CreateEmbedFooter::new(
                format!("👍 {}", res_last.ups)
            ))],
        false
    );

    Ok(())
}
